Validate ID Token Issuer

furti opened the pull request:
When the issuer is set in the provider metadata, we validate the iss
field of the ID Token against it.

The OpenID Connect Specification says this must always be validated.
But this would be a breaking change for applications configured other
than with ClientRegistrations.fromOidcIssuerLocation(issuer). This will
be done later with #8326

fixes gh-8321

### Running gradlew
I tried to run ```./gradlew build``` but 3 Tests failed in classes I did not even touch. But the tests for the class I changed passed, so I think the changes should be fine.

If you want me to give the build another try, please let me know.

### CLA
I signed the CLA.

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java:
```diff
@@ -143,6 +146,13 @@ public void setClock(Clock clock) {
 
+	private boolean issuerMatchesMetadata(Jwt idToken) {
+		String metadataIssuer = (String) clientRegistration.getProviderDetails().getConfigurationMetadata()
+				.get("issuer");
```
The issuer claim is `iss`. You can use `JwtClaimNames.ISS`.

mengelbrecht replied:
Are you sure? The claim is called `issuer` in the provider metadata and `iss` in the id token (https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata).

jgrandja replied:
Of course, my mistake. Thanks for correcting me @mengelbrecht !

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java:
```diff
@@ -68,7 +69,9 @@ public OAuth2TokenValidatorResult validate(Jwt idToken) {
 		// 2. The Issuer Identifier for the OpenID Provider (which is typically obtained during Discovery)
 		// MUST exactly match the value of the iss (issuer) Claim.
-		// TODO Depends on gh-4413
+		if (!issuerMatchesMetadata(idToken)) {
```
Can you move the validation check here and remove the method.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java:
```diff
@@ -143,6 +146,13 @@ public void setClock(Clock clock) {
 	}
 
+	private boolean issuerMatchesMetadata(Jwt idToken) {
+		String metadataIssuer = (String) clientRegistration.getProviderDetails().getConfigurationMetadata()
```
We prefix `this.` for private members -> `this.clientRegistration`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java:
```diff
@@ -92,6 +92,35 @@ public void validateWhenIssuerNullThenHasErrors() {
+		 * the validation must fail
+		 */
+		Map<String, Object> configurationMetadata = new HashMap<>();
+		configurationMetadata.put("issuer", "https://issuer.somethingelse.com");
```
Use `JwtClaimNames.ISS`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java:
```diff
@@ -92,6 +92,35 @@ public void validateWhenIssuerNullThenHasErrors() {
+		 * the validation must fail
+		 */
+		Map<String, Object> configurationMetadata = new HashMap<>();
+		configurationMetadata.put("issuer", "https://issuer.example.com");
```
Use `JwtClaimNames.ISS`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java:
```diff
@@ -92,6 +92,35 @@ public void validateWhenIssuerNullThenHasErrors() {
+	public void validateWhenMetadataIssuerMatchThenNoErrors() {
+		/*
+		 * When the issuer is set in the provider metadata, and it does not match the issuer in the ID Token,
+		 * the validation must fail
```
validation must fail -> validation will pass

furti commented:
@jgrandja I implemented your changes and pushed the commit again.

I moved the check inline, added the "this." prefix and fixed the wrong javadoc comment in the Test (copy paste error).
I ignored the "issuer" -> "iss" comments, as @mengelbrecht pointed out, the claim is called "issuer" in the metadata.
